rs/src/management/http_client.rs

```diff
@@ -0,0 +1,560 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+use reqwest::{
+    header::{HeaderValue, AUTHORIZATION, CONTENT_TYPE},
+    Client, Method, Request,
+};
+use serde::{de::DeserializeOwned, Deserialize, Serialize};
+use url::Url;
+
+use crate::contracts::{
+    env_production, Tunnel, TunnelConnectionMode, TunnelEndpoint, TunnelEnvironment, TunnelPort,
+};
+
+use super::{
+    Authorization, HttpError, HttpResult, ResponseError, TunnelLocator, TunnelRequestOptions,
+};
+
+pub struct TunnelManagementClient {
+    client: Client,
+    authorization: Option<HeaderValue>,
```

Comment:
Other languages have an async callback for getting the auth header, in case a token needs to be refreshed or something like that. I don't know what would be the best way to handle that in Rust.

Reply:
Async is rust is always fun, but I implemented that 😬  Consumers can now pass something like I do in tests:

```rs
        c.authorization_fn(|| async {
            let token = match env::var("TUNNEL_TEST_AAD_TOKEN") {
                Ok(value) => value,
                _ => do_device_code_flow(&reqwest::Client::new()).await,
            };

            env::set_var("TUNNEL_TEST_AAD_TOKEN", &token);
            Ok(Authorization::Bearer(token))
        });
```